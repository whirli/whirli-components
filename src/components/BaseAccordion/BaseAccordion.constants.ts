export const PropKeys = {
  STATE: 'state',
  NAME: 'name',
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.STATE]: ['open', 'closed'],
};
