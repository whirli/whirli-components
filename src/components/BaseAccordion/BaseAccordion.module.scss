@import '@whirli-components/styles/Breakpoints.module.scss';

// default styles
.accordion__trigger {
  background: none;
  border: none;
  box-shadow: none;
  color: inherit;
  cursor: pointer;
}

.accordion__content {
  overflow: hidden;
  transition: max-height ease-in-out 0.5s;
}

[data-state="open"] .accordion__content {
  max-height: auto;
  max-height: var(--accordion-height);
}

[data-state="closed"] .accordion__content {
  max-height: 0;
}

// breakpoint styles
$package-breakpoint-styles: ();
