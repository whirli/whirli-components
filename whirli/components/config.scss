@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}