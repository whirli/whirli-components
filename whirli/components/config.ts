import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
