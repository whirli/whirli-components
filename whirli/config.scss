@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}