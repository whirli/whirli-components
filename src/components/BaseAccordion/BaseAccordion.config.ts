// @ts-ignore
import { PropType } from 'vue';
import * as PropHelpers from '../../helpers/props';
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli-local/components/BaseAccordion/BaseAccordion.config';

// Package config
import { PropKeys } from './BaseAccordion.constants';
import type { PropState } from './BaseAccordion.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.STATE]: {
    generateBreakpoints: true,
    classes: {
      closed: 'state-closed',
      open: 'state-open',
      ...config.styles[PropKeys.STATE]?.classes,
    } as Record<PropState, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.NAME]: {
    required: false,
    type: String as PropType<string>,
  },
  [PropKeys.STATE]: {
    required: false,
    type: [String, Object] as PropType<PropState>,
    default: 'closed',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.STATE),
  },
  ...config.props,
};
