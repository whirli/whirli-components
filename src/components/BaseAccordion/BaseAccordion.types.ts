import { PropKeys, PropValues } from './BaseAccordion.constants';

const propState: string[] = PropValues[PropKeys.STATE];

export type PropState = typeof propState[number];
