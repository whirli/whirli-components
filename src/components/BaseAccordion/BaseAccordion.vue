<template>
  <div :class="wrapperClasses">
    <button
      :class="triggerClasses"
      @click="setAccordionToggleState()"
      :aria-controls="bodyId"
      :aria-expanded="isAccordionOpen"
      :id="headerId"
    >
      <slot name="title" />
    </button>
    <div
      :id="bodyId"
      ref="accordionContentWrapper"
      :class="contentClasses"
      :style="`--accordion-height: ${accordionMaxHeight}`"
      :aria-labelledby="headerId"
      :aria-hidden="!isAccordionOpen"
    >
      <div ref="accordionContent"><slot name="content" /></div>
    </div>
  </div>
</template>

<script setup="props" lang="ts">
// Vue
import { defineProps, computed, ref, onMounted, onUnmounted } from 'vue';
import type { ComputedRef, Ref } from 'vue';

// Types
import {
  ComponentStyles as ComponentStylesInterface,
  ComponentProps as ComponentPropsInterface,
} from '../../@types/components';

// Styles
import styles from '@whirli-local/components/BaseAccordion/BaseAccordion.module.scss';

// Data
import { ConfigStyles, ConfigProps } from './BaseAccordion.config';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

const props: ComponentPropsInterface = defineProps(ConfigProps);

const TOGGLE_STATES: Record<string, string> = {
  DEFAULT: 'toggle-default',
  ALT: 'toggle-alt',
};

const accordionContentWrapper: Ref<HTMLElement> = ref({} as HTMLElement);
const accordionContent: Ref<HTMLElement> = ref({} as HTMLElement);
const accordionMaxHeight: Ref<string> = ref('');
const accordionToggleState: Ref<string> = ref(TOGGLE_STATES.DEFAULT);
const isAccordionOpen: Ref<boolean> = ref(false);
const headerId = `${props.name}-header-id`;
const bodyId = `${props.name}-content-id`;

// Classes
import useClasses from '../../@use/class';
const { makeClasses } = useClasses();
const wrapperClasses: ComputedRef<string[]> = computed(() => [
  styles.accordion,
  styles[accordionToggleState.value],
  ...makeClasses(ComponentStyles, props, styles),
]);
const triggerClasses: string[] = [styles['accordion__trigger']];
const contentClasses: string[] = [styles['accordion__content']];

function updateAccordionHeight(): void {
  accordionMaxHeight.value = `${accordionContent.value.offsetHeight}px`;
}

function setAccordionToggleState(): void {
  accordionToggleState.value =
    accordionToggleState.value === TOGGLE_STATES.DEFAULT ? TOGGLE_STATES.ALT : TOGGLE_STATES.DEFAULT;
  setAccordionOpenState();
}

function setAccordionOpenState(): void {
  // @todo - Revisit this function
  setTimeout(() => {
    isAccordionOpen.value = accordionContentWrapper.value.clientHeight > 0;
  }, 1000);
}

function windowResized(): void {
  accordionToggleState.value = TOGGLE_STATES.DEFAULT;
  setAccordionOpenState();
}

onMounted(() => {
  updateAccordionHeight();
  setAccordionOpenState();
  window.addEventListener('resize', windowResized);
});

onUnmounted(() => {
  window.removeEventListener('resize', windowResized);
});
</script>
