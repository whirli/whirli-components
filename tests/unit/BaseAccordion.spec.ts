import { mount } from '@vue/test-utils';

import { propReturnsClass, propReturnsBreakpointClasses } from '../helpers';

import BaseAccordion from '@/components/BaseAccordion/BaseAccordion.vue';
import { PropKeys } from '@/components/BaseAccordion/BaseAccordion.constants';
import { ConfigStyles } from '@/components/BaseAccordion/BaseAccordion.config';

let wrapper: any;
const title = 'Title';
const content = 'Hello World';

describe('BaseAccordion.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = mount(BaseAccordion, {
      slots: {
        title,
        content,
      },
      props: {
        name: 'accordion',
        state: 'open',
      },
    });
    wrapper.vm.setAccordionToggleState = jest.fn();
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct title', () => {
    expect(wrapper.find('#accordion-header-id').text()).toBe(title);
  });

  it('renders the correct content', () => {
    expect(wrapper.find('#accordion-content-id').text()).toBe(content);
  });

  it('is the correct state', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.STATE, 'state-open')).toBe(true);
  });

  it('calls setAccordionToggleState on Button click', async () => {
    wrapper.find('#accordion-header-id').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.setAccordionToggleState).toHaveBeenCalled();
  });
});

describe('BaseAccordion.vue - Breakpoint classes', () => {
  beforeEach(() => {
    wrapper = mount(BaseAccordion, {
      slots: {
        title,
        content,
      },
      props: {
        name: 'accordion',
        state: { default: 'open', tablet: 'closed' },
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct title', () => {
    expect(wrapper.find('#accordion-header-id').text()).toBe(title);
  });

  it('renders the correct content', () => {
    expect(wrapper.find('#accordion-content-id').text()).toBe(content);
  });

  it('is the correct breakpoint states', () => {
    expect(
      propReturnsBreakpointClasses(wrapper, ConfigStyles, PropKeys.STATE, {
        default: 'state-open',
        tablet: 'state-closed',
      })
    ).toBe(true);
  });
});
