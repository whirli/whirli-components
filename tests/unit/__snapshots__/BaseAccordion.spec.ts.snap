// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseAccordion.vue - Breakpoint classes renders the component 1`] = `
<div
  class=""
>
  <button
    aria-controls="accordion-content-id"
    aria-expanded="false"
    class=""
    id="accordion-header-id"
  >
    
    Title
    
  </button>
  <div
    aria-hidden="true"
    aria-labelledby="accordion-header-id"
    class=""
    id="accordion-content-id"
    style="--accordion-height: 0px;"
  >
    <div>
      
      Hello World
      
    </div>
  </div>
</div>
`;

exports[`BaseAccordion.vue - Single classes renders the component 1`] = `
<div
  class=""
>
  <button
    aria-controls="accordion-content-id"
    aria-expanded="false"
    class=""
    id="accordion-header-id"
  >
    
    Title
    
  </button>
  <div
    aria-hidden="true"
    aria-labelledby="accordion-header-id"
    class=""
    id="accordion-content-id"
    style="--accordion-height: 0px;"
  >
    <div>
      
      Hello World
      
    </div>
  </div>
</div>
`;
